src/lib/optimizer/strategy/predicate_reordering_rule.cpp

```diff
@@ -40,27 +46,35 @@ void PredicateReorderingRule::apply_to(const std::shared_ptr<AbstractLQPNode>& n
      * Continue rule in deepest input
      */
     if (predicate_nodes.size() > 1) {
+      const auto input = predicate_nodes.back()->left_input();
       _reorder_predicates(predicate_nodes);
-      _apply_to_inputs(predicate_nodes.back());
-      return;
+      apply_to(input);
     }
   }
 
   _apply_to_inputs(node);
 }
 
-void PredicateReorderingRule::_reorder_predicates(std::vector<std::shared_ptr<AbstractLQPNode>>& predicates) const {
+void PredicateReorderingRule::_reorder_predicates(
+    const std::vector<std::shared_ptr<AbstractLQPNode>>& predicates) const {
   // Store original input and output
   auto input = predicates.back()->left_input();
   const auto outputs = predicates.front()->outputs();
   const auto input_sides = predicates.front()->get_input_sides();
 
-  const auto sort_predicate = [&](auto& left, auto& right) {
-    return left->derive_statistics_from(input)->row_count() > right->derive_statistics_from(input)->row_count();
-  };
-
-  if (std::is_sorted(predicates.begin(), predicates.end(), sort_predicate)) {
-    return;
+  // Setup cardinality estimation cache so that the statistics of `input` (which might be a big plan) do not need to
+  // be determined repeatedly.
+  // We do NOT guarantee bottom-up construction, but we can guarantee that `input` is a "vertex" below which we will not
+  // change the plan. Thus we construct a dummy-JoinGraph and allow the CardinalityEstimator to use it.
```

Comment:
This sounds like a hack that makes the cache do something that it would usually just do for join graphs. I do not yet understand why this is necessary and what this accomplishes.

Reply:
Assume you have a big, big subplan with root node x and on top of the suplan you want to order 3 predicates (a, b, c). The PredicateReorderingRule considers the cardinalities of x->a, x->b and x->c and chooses the order of a, b, c based on those cardinalities. Now that's three cardinality estimations where the entirety of x needs to be estimated because how is the estimator to know that you changed a to b, but everything below that is constant?

I could think of no way more practical than saying: "x and everything below it is constant" than doing the same thing the JoinOrderingRule does: Marking x as a "vertex".

Reply:
I think I'm missing a comment here that explains how using a JoinGraph changes things.

Reply:
Tried again, at loss whether this makes it better:

```
  // Setup cardinality estimation cache so that the statistics of `input` (which might be a big plan) do not need to
  // be determined repeatedly. For this, we hijack the `guarantee_join_graph()`-guarantee and via it promise the
  // CardinalityEstimator that we will not change the LQP below the `input` node by marking it as a "vertex". 
  // This allows the CardinalityEstimator to compute the statistics of `input` once, chache them and then re-use them.
```